Terraform crashed: interface {} is string, not float64

### Terraform Version

```shell
1.5.2
```


### Terraform Configuration Files

```terraform
...terraform config...
```


### Debug Output

```
!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

Terraform crashed! This is always indicative of a bug within Terraform.
Please report the crash with Terraform[1] so that we can fix this.

When reporting bugs, please include your terraform version, the stack trace
shown below, and any additional information which may help replicate the issue.

[1]: https://github.com/hashicorp/terraform/issues

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!


Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
interface conversion: interface {} is string, not float64
goroutine 37 [running]:
runtime/debug.Stack()
        /Users/runner/hostedtoolcache/go/1.20.0/x64/src/runtime/debug/stack.go:24 +0x65
runtime/debug.PrintStack()
        /Users/runner/hostedtoolcache/go/1.20.0/x64/src/runtime/debug/stack.go:16 +0x19
github.com/hashicorp/terraform/internal/logging.PanicHandler()
        /Users/runner/work/terraform/terraform/internal/logging/panic.go:58 +0x153
panic({0x304bf80, 0xc002452d80})
        /Users/runner/hostedtoolcache/go/1.20.0/x64/src/runtime/panic.go:884 +0x213
github.com/hashicorp/terraform/internal/command/jsonformat/structured/attribute_path.(*PathMatcher).GetChildWithIndex(0xc00013be60, 0x0)
        /Users/runner/work/terraform/terraform/internal/command/jsonformat/structured/attribute_path/matcher.go:176 +0x2cb
github.com/hashicorp/terraform/internal/command/jsonformat/structured.ChangeSlice.GetChild({{0xc002bd69e0, 0x1, 0x1}, {0xc002bd6da0, 0x1, 0x1}, {0x0, 0x0, 0x0}, {0xc002bd6f30, ...}, ...}, ...)
        /Users/runner/work/terraform/terraform/internal/command/jsonformat/structured/slice.go:75 +0x252
github.com/hashicorp/terraform/internal/command/jsonformat/differ.processNestedList({0x1, 0x1, {0x2ed0720, 0xc0019cb980}, {0x2ed0720, 0xc0019cbb60}, {0x0, 0x0}, {0x2ed0720, 0xc0019cbd58}, ...}, ...)
        /Users/runner/work/terraform/terraform/internal/command/jsonformat/differ/list.go:83 +0x13b
github.com/hashicorp/terraform/internal/command/jsonformat/differ.computeBlockDiffsAsList({0x1, 0x1, {0x2ed0720, 0xc0019cb980}, {0x2ed0720, 0xc0019cbb60}, {0x0, 0x0}, {0x2ed0720, 0xc0019cbd58}, ...}, ...)
        /Users/runner/work/terraform/terraform/internal/command/jsonformat/differ/list.go:72 +0x145
github.com/hashicorp/terraform/internal/command/jsonformat/differ.ComputeDiffForBlock({0x0, 0x0, {0x3007cc0, 0xc0022e2c90}, {0x3007cc0, 0xc0022e2f00}, {0x3007cc0, 0xc0022e3140}, {0x3007cc0, 0xc0022e3170}, ...}, ...)
        /Users/runner/work/terraform/terraform/internal/command/jsonformat/differ/block.go:92 +0xf98
github.com/hashicorp/terraform/internal/command/jsonformat.precomputeDiffs({{0x33dc6ed, 0x3}, 0xc001bf5fb0, {0xc0006cf000, 0x13, 0x22}, {0xc0012fc800, 0xa, 0x11}, {0xc0019fd680, ...}, ...}, ...)
        /Users/runner/work/terraform/terraform/internal/command/jsonformat/diff.go:52 +0x3b8
github.com/hashicorp/terraform/internal/command/jsonformat.Plan.renderHuman({{0x33dc6ed, 0x3}, 0xc001bf5fb0, {0xc0006cf000, 0x13, 0x22}, {0xc0012fc800, 0xa, 0x11}, {0xc0019fd680, ...}, ...}, ...)
        /Users/runner/work/terraform/terraform/internal/command/jsonformat/plan.go:64 +0xa5
github.com/hashicorp/terraform/internal/command/jsonformat.Renderer.RenderHumanPlan({0xc000120d68?, 0xc0002a6cc0?, 0x17?}, {{0x33dc6ed, 0x3}, 0xc001bf5fb0, {0xc0006cf000, 0x13, 0x22}, {0xc0012fc800, ...}, ...}, ...)
        /Users/runner/work/terraform/terraform/internal/command/jsonformat/renderer.go:92 +0x1ef
github.com/hashicorp/terraform/internal/command/views.(*OperationHuman).Plan(0xc000ac0790, 0xc0010983c0, 0x0?)
        /Users/runner/work/terraform/terraform/internal/command/views/operation.go:126 +0x31c
github.com/hashicorp/terraform/internal/backend/local.(*Local).opApply(0x0?, {0x39deb38, 0xc0008343c0}, {0x39deb38, 0xc000834410}, 0xc000aba5a0, 0xc000aaed00)
        /Users/runner/work/terraform/terraform/internal/backend/local/backend_apply.go:117 +0x996
github.com/hashicorp/terraform/internal/backend/local.(*Local).Operation.func1()
        /Users/runner/work/terraform/terraform/internal/backend/local/backend.go:326 +0xc3
created by github.com/hashicorp/terraform/internal/backend/local.(*Local).Operation
        /Users/runner/work/terraform/terraform/internal/backend/local/backend.go:319 +0x44a
make: *** [apply] Error 11
```

### Expected Behavior

I expected terraform apply successfully

### Actual Behavior

terraform crashed

### Steps to Reproduce

1. `terraform init`
2. `terraform apply`

### Additional Context

_No response_

### References

_No response_

--- a/internal/command/jsonformat/differ/differ_test.go
+++ b/internal/command/jsonformat/differ/differ_test.go
@@ -2742,6 +2742,115 @@ func TestSpecificCases(t *testing.T) {
 				}, plans.Create, false),
 			}, nil, nil, nil, nil, plans.Create, false),
 		},
+
+		// The following tests are from issue 33472. Basically Terraform allows
+		// callers to treat numbers as strings in references and expects us
+		// to coerce the strings into numbers. For example the following are
+		// equivalent.
+		//    - test_resource.resource.list[0].attribute
+		//    - test_resource.resource.list["0"].attribute
+		//
+		// We need our attribute_path package (used within the ReplacePaths and
+		// RelevantAttributes fields) to handle coercing strings into numbers
+		// when it's expected.
+
+		"issues/33472/expected": {
+			input: structured.Change{
+				Before: map[string]interface{}{
+					"list": []interface{}{
+						map[string]interface{}{
+							"number": -1,
+						},
+					},
+				},
+				After: map[string]interface{}{
+					"list": []interface{}{
+						map[string]interface{}{
+							"number": 2,
+						},
+					},
+				},
+				Unknown:         false,
+				BeforeSensitive: false,
+				AfterSensitive:  false,
+				ReplacePaths:    attribute_path.Empty(false),
+				RelevantAttributes: &attribute_path.PathMatcher{
+					Propagate: true,
+					Paths: [][]interface{}{
+						{
+							"list",
+							0.0, // This is normal and expected so easy case.
+							"number",
+						},
+					},
+				},
+			},
+			block: &jsonprovider.Block{
+				Attributes: map[string]*jsonprovider.Attribute{
+					"list": {
+						AttributeType: unmarshalType(t, cty.List(cty.Object(map[string]cty.Type{
+							"number": cty.Number,
+						}))),
+					},
+				},
+			},
+			validate: renderers.ValidateBlock(map[string]renderers.ValidateDiffFunction{
+				"list": renderers.ValidateList([]renderers.ValidateDiffFunction{
+					renderers.ValidateObject(map[string]renderers.ValidateDiffFunction{
+						"number": renderers.ValidatePrimitive(-1, 2, plans.Update, false),
+					}, plans.Update, false),
+				}, plans.Update, false),
+			}, nil, nil, nil, nil, plans.Update, false),
+		},
+
+		"issues/33472/coerce": {
+			input: structured.Change{
+				Before: map[string]interface{}{
+					"list": []interface{}{
+						map[string]interface{}{
+							"number": -1,
+						},
+					},
+				},
+				After: map[string]interface{}{
+					"list": []interface{}{
+						map[string]interface{}{
+							"number": 2,
+						},
+					},
+				},
+				Unknown:         false,
+				BeforeSensitive: false,
+				AfterSensitive:  false,
+				ReplacePaths:    attribute_path.Empty(false),
+				RelevantAttributes: &attribute_path.PathMatcher{
+					Propagate: true,
+					Paths: [][]interface{}{
+						{
+							"list",
+							"0", // Difficult but allowed, we need to handle this.
+							"number",
+						},
+					},
+				},
+			},
+			block: &jsonprovider.Block{
+				Attributes: map[string]*jsonprovider.Attribute{
+					"list": {
+						AttributeType: unmarshalType(t, cty.List(cty.Object(map[string]cty.Type{
+							"number": cty.Number,
+						}))),
+					},
+				},
+			},
+			validate: renderers.ValidateBlock(map[string]renderers.ValidateDiffFunction{
+				"list": renderers.ValidateList([]renderers.ValidateDiffFunction{
+					renderers.ValidateObject(map[string]renderers.ValidateDiffFunction{
+						"number": renderers.ValidatePrimitive(-1, 2, plans.Update, false),
+					}, plans.Update, false),
+				}, plans.Update, false),
+			}, nil, nil, nil, nil, plans.Update, false),
+		},
 	}
 	for name, tc := range tcs {
 		t.Run(name, func(t *testing.T) {

--- a/internal/command/jsonformat/structured/attribute_path/matcher.go
+++ b/internal/command/jsonformat/structured/attribute_path/matcher.go
@@ -3,7 +3,11 @@
 
 package attribute_path
 
-import "encoding/json"
+import (
+	"encoding/json"
+	"fmt"
+	"strconv"
+)
 
 // Matcher provides an interface for stepping through changes following an
 // attribute path.
@@ -173,8 +177,32 @@ func (p *PathMatcher) GetChildWithIndex(index int) Matcher {
 			continue
 		}
 
-		if int(path[0].(float64)) == index {
-			child.Paths = append(child.Paths, path[1:])
+		// Terraform actually allows user to provide strings into indexes as
+		// long as the string can be interpreted into a number. For example, the
+		// following are equivalent and we need to support them.
+		//    - test_resource.resource.list[0].attribute
+		//    - test_resource.resource.list["0"].attribute
+		//
+		// Note, that Terraform will raise a validation error if the string
+		// can't be coerced into a number, so we will panic here if anything
+		// goes wrong safe in the knowledge the validation should stop this from
+		// happening.
+
+		switch val := path[0].(type) {
+		case float64:
+			if int(path[0].(float64)) == index {
+				child.Paths = append(child.Paths, path[1:])
+			}
+		case string:
+			f, err := strconv.ParseFloat(val, 64)
+			if err != nil {
+				panic(fmt.Errorf("found invalid type within path (%v:%T), the validation shouldn't have allowed this to happen; this is a bug in Terraform, please report it", val, val))
+			}
+			if int(f) == index {
+				child.Paths = append(child.Paths, path[1:])
+			}
+		default:
+			panic(fmt.Errorf("found invalid type within path (%v:%T), the validation shouldn't have allowed this to happen; this is a bug in Terraform, please report it", val, val))
 		}
 	}
 	return child

